File: motor3d 2.0.ConvertedToC#/Segmento3D.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using System.Math;

namespace Motor3D.Espacio3D
{
	public class Segmento3D : ObjetoGeometrico3D
	{

		private Punto3D mExtremoInicial;
		private Punto3D mExtremoFinal;
		private Recta3D mRecta;
		private double mLongitud;

		private Caja3D mCaja;
		public Punto3D ExtremoInicial {
			get { return mExtremoInicial; }
			set {
				mExtremoInicial = value;
				RecalcularDatos();
			}
		}

		public Punto3D ExtremoFinal {
			get { return mExtremoFinal; }
			set {
				mExtremoFinal = value;
				RecalcularDatos();
			}
		}

		public Recta3D Recta {
			get { return mRecta; }
		}

		public Caja3D Caja {
			get { return mCaja; }
		}

		public double Longitud {
			get { return mLongitud; }
		}

		public Punto3D PuntoMedio {
			get { return new Punto3D((mExtremoInicial.X + mExtremoFinal.X) / 2, (mExtremoInicial.Y + mExtremoFinal.Y) / 2, (mExtremoInicial.Z + mExtremoFinal.Z) / 2); }
		}

		public Segmento3D(Punto3D P1, Punto3D P2)
		{
			mExtremoInicial = P1;
			mExtremoFinal = P2;

			mLongitud = Math.Sqrt((Math.Pow((P1.X + P2.X), 2)) + (Math.Pow((P1.Y + P2.Y), 2)));
			mRecta = new Recta3D(P1, P2);

			mCaja = new Caja3D(P1, new Vector3D(P2.X - P1.X, P2.Y - P1.Y, P2.Z - P1.Z));
		}

		private void RecalcularDatos()
		{
			mRecta = new Recta3D(mExtremoInicial, mExtremoFinal);
			mCaja = new Caja3D(mExtremoInicial, new Vector3D(mExtremoFinal.X - mExtremoInicial.X, mExtremoFinal.Y - mExtremoInicial.Y, mExtremoFinal.Z - mExtremoInicial.Z));
			mLongitud = Math.Sqrt((Math.Pow((mExtremoInicial.X + mExtremoFinal.X), 2)) + (Math.Pow((mExtremoInicial.Y + mExtremoFinal.Y), 2)));
		}

		public bool Pertenece(Punto3D Punto)
		{
			return Pertenece(this, Punto);
		}

		public bool Pertenece(Segmento3D Segmento, Punto3D Punto)
		{
			return mCaja.Pertenece(Punto) && mRecta.Pertenece(Punto);
		}

		public override string ToString()
		{
			return "{Segmento. Inicio=" + mExtremoInicial.ToString() + ",Fin=" + mExtremoFinal.ToString() + "}";
		}
	}
}

File: motor3d 2.0/Segmento3D.vb
Imports System.Math

Namespace Espacio3D
    Public Class Segmento3D
        Inherits ObjetoGeometrico3D

        Private mExtremoInicial As Punto3D
        Private mExtremoFinal As Punto3D
        Private mRecta As Recta3D
        Private mLongitud As Double
        Private mAABB As AABB3D

        Public Property ExtremoInicial() As Punto3D
            Get
                Return mExtremoInicial
            End Get
            Set(ByVal value As Punto3D)
                mExtremoInicial = value
                RecalcularDatos()
            End Set
        End Property

        Public Property ExtremoFinal() As Punto3D
            Get
                Return mExtremoFinal
            End Get
            Set(ByVal value As Punto3D)
                mExtremoFinal = value
                RecalcularDatos()
            End Set
        End Property

        Public ReadOnly Property Recta() As Recta3D
            Get
                Return mRecta
            End Get
        End Property

        Public ReadOnly Property AABB() As AABB3D
            Get
                Return mAABB
            End Get
        End Property

        Public ReadOnly Property Longitud() As Double
            Get
                Return mLongitud
            End Get
        End Property

        Public ReadOnly Property PuntoMedio() As Punto3D
            Get
                Return New Punto3D((mExtremoInicial.X + mExtremoFinal.X) / 2, (mExtremoInicial.Y + mExtremoFinal.Y) / 2, (mExtremoInicial.Z + mExtremoFinal.Z) / 2)
            End Get
        End Property

        Public Sub New(ByVal P1 As Punto3D, ByVal P2 As Punto3D)
            mExtremoInicial = P1
            mExtremoFinal = P2

            mLongitud = Sqrt(((P1.X + P2.X) ^ 2) + ((P1.Y + P2.Y) ^ 2))
            mRecta = New Recta3D(P1, P2)

            mAABB = New AABB3D(P1, New Vector3D(P2.X - P1.X, P2.Y - P1.Y, P2.Z - P1.Z))
        End Sub

        Private Sub RecalcularDatos()
            mRecta = New Recta3D(mExtremoInicial, mExtremoFinal)
            mAABB = New AABB3D(mExtremoInicial, New Vector3D(mExtremoFinal.X - mExtremoInicial.X, mExtremoFinal.Y - mExtremoInicial.Y, mExtremoFinal.Z - mExtremoInicial.Z))
            mLongitud = Sqrt(((mExtremoInicial.X + mExtremoFinal.X) ^ 2) + ((mExtremoInicial.Y + mExtremoFinal.Y) ^ 2))
        End Sub

        Public Function Pertenece(ByVal Punto As Punto3D) As Boolean
            Return Pertenece(Me, Punto)
        End Function

        Public Function Pertenece(ByVal Segmento As Segmento3D, ByVal Punto As Punto3D) As Boolean
            Return mAABB.Pertenece(Punto) AndAlso mRecta.Pertenece(Punto)
        End Function

        Public Overrides Function ToString() As String
            Return "{Segmento. Inicio=" & mExtremoInicial.ToString & ",Fin=" & mExtremoFinal.ToString & "}"
        End Function
    End Class
End Namespace